**One-bin 2A + 5B <-> 4C + 3D, with 1st-order kinetics for each species,
taken to equilibrium**

Diffusion not applicable (just 1 bin)

LAST REVISED: July 2, 2022

In [1]:
import set_path
set_path.add_ancestor_dir_to_syspath(3)  # The number of levels to go up 
                                         # to reach the project's home, from the folder containing this notebook

Added 'D:\Docs\- MY CODE\BioSimulations\life123-Win7' to sys.path


In [2]:
from modules.chemicals.chemicals import Chemicals as chem
from modules.reactions.reactions import Reactions
from life_1D.bio_sim_1d import BioSim1D as bio

from modules.html_log.html_log import HtmlLog as log

In [3]:
# Initialize the system
chem_data = chem(names=["A", "B", "C", "D"])     # NOTE: Diffusion not applicable (just 1 bin)

rxn = Reactions(chem_data)

# Reaction 2A + 5B <-> 4C + 3D , with 1st-order kinetics for each species
rxn.add_reaction(reactants=[(2,"A") , (5,"B")], products=[(4,"C") , (3,"D")],
                 forward_rate=5., reverse_rate=2.)

bio.initialize_system(n_bins=1, chem_data=chem_data, reactions=rxn)

bio.set_all_uniform_concentrations( [4., 7., 5., 2.] )

bio.describe_state()

1 bins and 4 species:

  Species 0 (A). Diff rate: NOT SET. Conc:  [4.]
  Species 1 (B). Diff rate: NOT SET. Conc:  [7.]
  Species 2 (C). Diff rate: NOT SET. Conc:  [5.]
  Species 3 (D). Diff rate: NOT SET. Conc:  [2.]


In [4]:
rxn.describe_reactions()

Number of reactions:  1
0: 2 A + 5 B <-> 4 C + 3 D  (Rf = 5.0 / Rb = 2.0)


In [5]:
rxn._internal_reactions_data()    # Low-level view of the reactions data

0: [(2, 0, 1), (5, 1, 1)] <-> [(4, 2, 1), (3, 3, 1)]   ; Fwd: 5.0 / Back: 2.0


In [6]:
# First step
bio.react(time_step=0.001, n_steps=1)
bio.describe_state()

1 bins and 4 species:

  Species 0 (A). Diff rate: NOT SET. Conc:  [3.76]
  Species 1 (B). Diff rate: NOT SET. Conc:  [6.4]
  Species 2 (C). Diff rate: NOT SET. Conc:  [5.48]
  Species 3 (D). Diff rate: NOT SET. Conc:  [2.36]


_Early in the reaction :_
[A] = 3.76 ,  [B] = 6.4 ,  [C] = 5.48 ,  [D] = 2.36

In [7]:
# Numerous more steps
bio.react(time_step=0.001, n_steps=40)

bio.describe_state()

1 bins and 4 species:

  Species 0 (A). Diff rate: NOT SET. Conc:  [2.80284552]
  Species 1 (B). Diff rate: NOT SET. Conc:  [4.00711381]
  Species 2 (C). Diff rate: NOT SET. Conc:  [7.39430896]
  Species 3 (D). Diff rate: NOT SET. Conc:  [3.79573172]


Consistent with the 5/2 ratio of forward/reverse rates (and the 1st order reactions),
the systems settles in the following equilibrium:  
[A] = 2.80284552 , [B] = 4.00711381 , [C] = 7.39430896 , [D] = 3.79573172

In [8]:
A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
C_eq = bio.bin_concentration(0, 2)
D_eq = bio.bin_concentration(0, 3)
print(f"Ratio of equilibrium concentrations ((C_eq * D_eq) / (A_eq * B_eq)) : {(C_eq * D_eq) / (A_eq * B_eq)}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations ((C_eq * D_eq) / (A_eq * B_eq)) : 2.498977017845165
Ratio of forward/reverse rates: 2.5
